### ISSUE TITLE
The docker Orb is too much of a closed box ! (for me, maybe for others)

### ISSUE THREAD

**Issue by cansjt** [jobteaser/circleci#37]
While bootstrapping new micro services, implemented in Python, I found the docker orb not that helpful. Part of it was my need to understand and adapt to the way things are done at JT, but some comes from the lack of adaptability of the Orb. Over time, I tried to get back to using the Orb as it is intended to be used. The latest, better polished iteration, of which is in [the jobflow-dags](https://github.com/jobteaser/jobflow-dags) repository. But while testing jobteaser/jobflow-dags#1 I stumbled upon one major issue that makes it definitely tough to use for me and I cannot find a way to work around, without duplicating and tweaking the Orb `build_image` in each of my CI workflows.

But maybe I need to rewind a bit, explain what I am doing. It is a bit long, but please bear with me ! :pray:

### Context

I like very tight feedback loops, meaning once I push something to github I want it built AFAP, to not wait, have time to switch (cognitive) context, before getting the outcome of the build. Part of my strategy to get that is exploit CircleCI's caching as much as possible. Mainly to cut lots of download time, whether it is downloading the occasional distribution package, or the tons of python packages that tend to get sucked in even the smallest project.

So I managed to download packages only once, each time the dependencies of a project change. I put those in a cache, that I can reuse whenever I need it. Either to run tests or build a docker image. This helps shave several minutes from most CI workflow runs. Only those after a dependency change take longer. Which in the long run saves me (and my colleagues) a lot of (valuable) time (and the company some peanuts).

 In the case of a building an image, I use BuildKit and commands such as `RUN --mount=type=bind,source=/...,target=/...` to be able to use the cache inside the image. That allows me to avoid using the `COPY`, which adds useless layers to the images. An alternate way to do this is use staged builds, and only pick the files needed from a previous stage. But that strategy is not very practical and lead me to do stuff I am clearly [not comfortable with nor proud of](https://github.com/jobteaser/tracking/blob/master/Dockerfile#L28) (it works, ... _until it won't_ :fearful:).

And I often ended up copying the content of this orb job to tweak it. Each time I tried to deviate a bit less, but I think there remain one issue that I can't work around.

When you build a docker image it needs a so called _context_, a working directory. That context generally is the root directory of you git working copy (e.g. `~/project` in the context of the CI). The important thing about that context is that AFAIK, you cannot fetch anything outside of it (~though writing this, I must admit, I may not have checked as thoroughly as I should have, if this is still true with BuildKit~ You cannot). Practically that means, you cannot mount or COPY anything outside of `~/project` while building an image in the CI.

But with the job in this orb, if I to access my, I have to restore it with a `pre-steps`. And if I want that cache to be accessible inside the `Dockerfile` (either with `COPY ...` or `RUN --mount ...`, doesn't matter), the cache needs to be inside the `~/project`. Problem, if I do that, then the job's `checkout` step will fail ! Indeed you cannot clone a repository inside a non empty directory...

What I would need, is a mean to be able to perform actions between the job's `checkout` step and the `build_image` one. In a general manner, the orb should work fine by default, but provide means to adapt to workflows other than the simple default one. Otherwise we end-up with a lot of code duplication (_see e.g._ [1][tr], [2][jfd], [3][jfa]). Each with its slight variations and accompanying bugs (E.g. I know I have workflows missing the trivy check, `latest` tagging... I also know I can easily reduce the amount of differences across our repositories, we have that in our backlog but still can not really use the orb job as it is. At least not at the expanse of sluggish build).

I am not sure how, and how much, of this would be applicable to other languages used in the company. For a language like go, where the artifact to install in the docker image is a single statically linked "native" binary, this is surely overkill. But I do think the strategy is fairly sound for dynamic languages. So I am confident this could benefit to services / application written in Ruby, JavaScript as well as Python.

### Additional notes:

Another issue with this orb is that it doesn't help in any ways to build your docker image locally, which is especially useful when bootstrapping a project. Something that should be common place in a micro-service architecture. 

While thinking about it, I had a few ideas (some bad, some good). I thought it could be useful to share them here:
- Splitting the job in smaller ones: does not work as each job in CircleCI runs in a different container, hence you can't have a job that sets docker up and one that builds the image, it simply does not work;
- Merging some commands could help, but probably is not enough: we could merge the `docker/setup` and `docker/login` commands. Maybe also merge the `docker/build_image` and `docker/trivy_check`. We probably don't want to merge the `docker/build_image` and `docker/push_image` ones (or not with some knobs to be able to not push images in some cases). This could help reduce the amount of code to copy when copying the build_image job for customization.
- Provide hooks, e.g. in the form of shell scripts on can add to its repository, _e.g._:
  ```bash
  # Around the docker build command:
  [ -x "docker/before-build.sh" ] && ./docker/pre-build.sh
  docker build ...
  [ -x "docker/post-build.sh" ] && ./docker/post-build.sh
  ```
[tr]: https://github.com/jobteaser/tracking/blob/master/.circleci/config.yml#L111
[jfd]: https://github.com/jobteaser/jobflow-dags/blob/8220d45ec3e616038ef50261a798d62d39729e5a/.circleci/config.yml#L80
[jfa]: https://github.com/jobteaser/jobflow/blob/chart-customization/.circleci/config.yml#L120

**Comment by galdorjt** [comment]
I believe I understand your point. The Docker orb (others too, but this one in particular) exports both the `build_image` job, which is the default way to build an image at JT, and a set of commands (used by the job) for the different steps.

This system was always meant to let people write their own jobs, either in their projects or in their own orbs; this is what we do in the go-service and rb-service orbs (e.g. https://github.com/jobteaser/circleci/blob/master/orbs/go-service/orb.yml#L62).

CircleCI orbs are really hard to develop and test, which is why we try to keep them really simple and avoid hooks (because with hooks, jobs may behave totally differently in a way which is really hard to debug).

I would strongly encourage you to start building your own orbs, using the commands exported from other orbs as building blocks. This is what we do ourselves. It will give you total control on your CI process.
